### PR TITLE
Fixes for quoted strings with protected " and / characters #11

### PR DESCRIPTION
**Pull request by lorton**
These unit tests show that issue #11 doesn't seem to be a general
problem

### CODE REVIEW

**Comment by lorton** [comment]
@mbeachy - I changed the indexed string parser so that when it knows it's parsing string blocks, it sets the buffer start +1 and buffer end -1.  This now returns unquoted strings, which matches behavior in other locations.  Please let me know if you think that's wrong for some reason.  The unit tests were previously explicitly confirming the old behavior, but I think that's more writing unit tests to match the previous behavior rather than writing unit tests that do what you want.

**Comment by speleo3** [comment]
@lorton - trimming the buffer by +1/-1 is not sufficient. It will not handle escaped quotes and backslashes.

In `test.mae`, change line 35 to:
```
  "Title with \" quote and \\ backslash"
```
In `ReaderTest.cpp`, change line 302 to:
```c++
    BOOST_REQUIRE_EQUAL(title, "Title with \" quote and \\ backslash");
```
Additional information: As expected, Maestro displays the title as:
```
Title with " quote and \ backslash
```

File: test/MaeParserTest.cpp [hunk 1]
```diff
@@ -993,7 +993,7 @@ BOOST_AUTO_TEST_CASE(TestQuotedStrings2)
 
     std::string z;
     z = get_string(ibb, 1);
-    BOOST_REQUIRE_EQUAL(z, std::string("\"bc \\\"\""));
+    BOOST_REQUIRE_EQUAL(z, std::string("bc \\\""));
```
Comment (speleo3) [inline test/MaeParserTest.cpp]: Should be:
```c++
    BOOST_REQUIRE_EQUAL(z, std::string("bc \""));
```

File: MaeParser.cpp [hunk 2]
```diff
@@ -492,8 +492,8 @@ void IndexedBlockBuffer::value(Buffer& buffer)
                     break;
                 }
                 ++buffer.current;
-                m_tokens_list.setTokenIndices(save - buffer.begin,
-                                              buffer.current - buffer.begin);
+                m_tokens_list.setTokenIndices(save - buffer.begin + 1,
```
Comment (speleo3) [inline MaeParser.cpp]: simply trimming by +1/-1 will not handle escaped quotes and backslashes

Comment (lorton) [reply MaeParser.cpp]: Dealing with quoted strings and slashes will definitely be a more substantial change we have to think about, as now we're talking about actually modifying the stored memory, which is something we try very hard not to do for performance reasons.  I'm not even sure if there *is* a way to do that without some substantial rewriting.

**Comment by lorton** [comment]
@speleo3 - OK, I think I figured out a reasonable fix here.  When parsing we still don't do any memory copying or editing of the quotes, but when we're actually putting the values into string vectors to be used, at that point we remove quotes and deprotect if necessary.  Added unit tests for both indexed and non-indexed values with quote-wrapping and protected characters.

File: test/MaeParserTest.cpp [hunk 3]
```diff
@@ -587,7 +587,7 @@ BOOST_AUTO_TEST_CASE(String)
         Buffer b(ss);
         whitespace(b);
         std::string s = parse_value<std::string>(b);
-        BOOST_REQUIRE_EQUAL(s, "Q\\ Z");
+        BOOST_REQUIRE_EQUAL(s, R"(Q\ Z)");
```
Comment (speleo3) [inline test/MaeParserTest.cpp]: If I load this string into maestro, it will be `Q Z`, not `Q\ Z`, so it looks like the mmshare implementation allows escaping any characters, not only `\` and `"`.

Comment (lorton) [reply test/MaeParserTest.cpp]: Ugh, but this makes no sense.  A protected b isn't just 'b', [it has a meaning](https://stackoverflow.com/questions/10220401/rules-for-c-string-literals-escape-character), if we strip the protection we're just ignoring that.  Even if they didn't have a meaning, The obvious example of that being true is the protected 'n', which should be interpreted as newline by whatever reads it.

I'm hesitating on trying to perfectly reproduce Maestro's behavior on edge cases where it behaves poorly.  I doubt we rely on this anywhere in Maestro.

Comment (speleo3) [reply test/MaeParserTest.cpp]: I can't disagree, but I think our reasoning is at a tie. You want to treat `"\n"` like `"\\n"`, I suggested to treat it like `"n"`, both cases have lost the original information. I still feel like I win because Maestro is on my side ;-)

Comment (lorton) [reply test/MaeParserTest.cpp]: Yeah, that's fair enough.  And if we're going to allow both parsers to exist in mmshare at once (which we're planning on doing), then I suppose they have to get as close to the exact same as possible - so it's either change m2io or change this.  I'll change this to just strip '\' naively.

**Comment by lorton** [comment]
@speleo3 - how are you testing what Maestro does, from investigating what it shows when you're in Maestro itself?  Infuriatingly it looks like that's even different than what happens when you just use your ultra-simple parse:
```
In [7]: st = structure.read_ct("test/test.mae")
In [8]: st.title
Out[8]: 'Title with p \\ " space'
In [10]: print(st.title)
Title with p \ " space
In [11]: st.atom[1].property['s_m_atom_name']
Out[11]: 'Does p " \\this work'
```

It looks like it strips all '\' except in the case of the double '\\'.  This makes sense I suppose, as it's relying on maestro's text display functions to do the stripping in that case.

**Comment by speleo3** [comment]
yes I was messing with the title field and loading it into Maestro.

> It looks like it strips all \ except in the case of the double \\.

Don't get fooled by looking at the Python string representation, it doesn't escape `"` when using single quotes for the string! The following three are identical:
```python
'Title with p \\ " space'
```
```python
"Title with p \\ \" space"
```
```python
r'Title with p \ " space'
```

**Comment by lorton** [comment]
For posterity - I've discussed with Thomas offline and we've decided we're going to add the exact unit tests we're adding here to mmct, to make sure we're preserving behavior exactly as it would be in mmct.  Once I'm done with that, I'll update this appropriately to make sure we have the same behavior.

**Comment by lorton** [comment]
GitHub seems to have gone loopy here...I've pushed a (hopefully final) version that goes back to simple stripping of escaped characters.  I'm also pushing into mmshare a unit test that confirms the behavior on that side.  That push is not showing up here, despite showing up if I look at the branch itself.  I'm going to close this and re-open a new pull request.
